### DISCUSSION TITLE
Custom Cross-Reference Type in a book: is there a way to associate each reference with a custom label?

### DISCUSSION THREAD

**Discussion by aborruso** [quarto-dev/quarto-cli#8034] [Feature Requests]
### Description

Hi,
I am creating an html book with quarto and I have added a Custom Cross-Reference Type, it's related to requirements:

```yaml
crossref:
  custom:
    - kind: float
      reference-prefix: Requisito
      key: req
```

One of it is in example the "requirement 4", and I refer to it using `@req-5`. 

````
::: {#req-5}

Nel caso in cui mettere a disposizione i propri dati secondo le indicazioni di cui sopra, per soddisfare richieste di apertura, comporti attività che vanno al di là della semplice manipolazione, e che, quindi, implicherebbero difficoltà sproporzionate, gli enti pubblici (cioè pubbliche amministrazioni e organismi di diritto pubblico) non hanno l'obbligo di adeguare i documenti o crearne nuovi o fornire estratti di documenti, fermi restando i limiti imposti dalla normativa in tema di protezione dei dati personali. In tal caso, attraverso un apposito provvedimento, l'ente titolare DEVE motivare le difficoltà sproporzionate indicando le attività sui dati che eccedono la semplice manipolazione.

:::
````

It works: in the rendering I have ` Requisito 4.1`, the prefix "Requisito" and chapter number first and then the order of the requirement (it is requirement number 5, but it is the only one I have written now).

I would like to have no automatic number, and instead of this a custom suffix label for this item, so that I can write "look at`@req-5`" and to have in output "look at Requisito 5". And `5` is the custom suffix I would like to have for requirement number five.

Is there something like this?

````
::: {#req-5 cr_label="5"}

Nel caso in cui mettere a disposizione i propri dati secondo le indicazioni di cui sopra, per soddisfare richieste di apertura, comporti attività che vanno al di là della semplice manipolazione, e che, quindi, implicherebbero difficoltà sproporzionate, gli enti pubblici (cioè pubbliche amministrazioni e organismi di diritto pubblico) non hanno l'obbligo di adeguare i documenti o crearne nuovi o fornire estratti di documenti, fermi restando i limiti imposti dalla normativa in tema di protezione dei dati personali. In tal caso, attraverso un apposito provvedimento, l'ente titolare DEVE motivare le difficoltà sproporzionate indicando le attività sui dati che eccedono la semplice manipolazione.

:::
````

Thank you

**Comment by cderv** [comment]
@cscheid can we do custom labels already ? I don't think so but I prefer to get your confirmation. Probably a new feature request if we don't have this.

**Reply by aborruso** [reply]
>  Probably a new feature request if we don't have this.

I think that would be something very useful.

**Reply by cscheid** [reply]
We don't have support for that.

**Comment by mcanouil** [comment]
Is it about labels or numbering?

<img width="726" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/e1f7d69f-c213-4cd6-b8b1-2c174989f672">


````qmd
---
title: "Quarto Playground"
format: html
crossref:
  custom:
    - kind: float
      reference-prefix: Something
      key: sg
---

This is a playground for Quarto.

## References

### Custom

::: {#sg-custom}
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Duis sagittis posuere ligula sit amet lacinia. Duis dignissim pellentesque magna, rhoncus congue sapien finibus mollis. Ut eu sem laoreet, vehicula ipsum in, convallis erat. Vestibulum magna sem, blandit pulvinar augue sit amet, auctor malesuada sapien. Nullam faucibus leo eget eros hendrerit, non laoreet ipsum lacinia. Curabitur cursus diam elit, non tempus ante volutpat a. Quisque hendrerit blandit purus non fringilla. Integer sit amet elit viverra ante dapibus semper. Vestibulum viverra rutrum enim, at luctus enim posuere eu. Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
:::

This is @sg-custom.

This is [my ref -@sg-custom].

This is [-@sg-custom].
````

**Reply by aborruso** [reply]
> Is it about labels or numbering?

Thank you @mcanouil.

As I wrote, it's about label: I would like to point to a cross reference, without then rendering of his automatic number and choose its label myself.

In some way to have a double `-`, and write

```
This is [my ref 5 --@sg-custom].
```

and get "This is [my ref 5](#)."

Or something like

```
::: {#sg-custom cr_label="5"}
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Duis sagittis posuere ligula sit amet lacinia. Duis dignissim pellentesque magna, rhoncus congue sapien finibus mollis. Ut eu sem laoreet, vehicula ipsum in, convallis erat. Vestibulum magna sem, blandit pulvinar augue sit amet, auctor malesuada sapien. Nullam faucibus leo eget eros hendrerit, non laoreet ipsum lacinia. Curabitur cursus diam elit, non tempus ante volutpat a. Quisque hendrerit blandit purus non fringilla. Integer sit amet elit viverra ante dapibus semper. Vestibulum viverra rutrum enim, at luctus enim posuere eu. Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
:::
```

and write

```
This is [my ref -@sg-custom].
```

to get "This is [my ref 5](#).".

**Reply by mcanouil** [reply]
You can already get the custom label: `This is [my custom label](#sg-custom).` (but without hover effect).

EDIT: to get back hover effect: `This is [my custom label](#sg-custom){.quarto-xref aria-expanded="false"}.`

**Reply by aborruso** [reply]
> You can already get the custom label: This is [my custom label](#sg-custom). (but without hover effect).

Thank you @mcanouil.

I'm a bad user of quarto, but not that bad 😇

If I'm not mistaken, your example is a normal link, it's not a cross reference. Then your example I think only works if the cross reference is on the same page. Or by writing a correct relative URL if it is on another page.

I have my [custom cross reference item](https://github.com/aborruso/linee-guida-opendata/blob/main/_quarto.yml#L77-L80) in [this page](https://aborruso.github.io/linee-guida-opendata/capitolo-4.html#req-5), it is the `@req-5`.

If I write in another page

```
# CR Label

See at @req-5

See at [a -@req-5]{.quarto-xref aria-expanded="false"}

This is [my custom label](#req-5){.quarto-xref aria-expanded="false"}.
```

I have [this](https://aborruso.github.io/linee-guida-opendata/debug/cr-label.html):

![image](https://github.com/quarto-dev/quarto-cli/assets/30607/388bf880-7d10-40a2-a90a-e6c721a5fec3)

I would like to be able, optionally, to point to that CR in another page, without writing a full relative URL (as `capitolo-4.html#req-5`), and much more conveniently from any page in any folder to be able to simply write `@req-5` - adding a suitable syntax - and have the link to it.

If it does not already exist, the example of the doble `-` seems interesting to me: remove twice, first remove the suffix, and then remove the number as well.
To write `See at [my label --@req-5]`, to get "See at [my label](#)", so that it points to that cross reference.

**Reply by mcanouil** [reply]
A cross-reference is a link.
Add the page name before the `#` and it should work everywhere.

**Reply by aborruso** [reply]
> A cross-reference is a link.
> Add the page name before the # and it should work everywhere.

I know, I wrote it. But if by chance I change the name of the page or its path nothing works anymore. Whereas for a cross reference its id is enough. I can change page names and path and it will continue to work.

**Reply by mcanouil** [reply]
I know, but there is currently no other way besides writing your own Lua filter to do this.

**Reply by aborruso** [reply]
> I know, but there is currently no other way besides writing your own Lua filter to do this.

Thank you, so it can't be done. Again, to me it sounds like a very interesting thing and as soon as I can I will add a feature request. I don't think I am able to write the script, but maybe I will try.

**Reply by aborruso** [reply]
> I know, but there is currently no other way besides writing your own Lua filter to do this.

[This is](https://lierdakil.github.io/pandoc-crossref/#reference-labels) interesting, but I have no idea on how to use it in quarto :(

**Reply by mcanouil** [reply]
it's a filter, use it as such and as described:
- https://github.com/lierdakil/pandoc-crossref
- https://quarto.org/docs/extensions/filters.html
